packages/next/build/plugins/collect-plugins.ts

```diff
@@ -0,0 +1,216 @@
+    }
+  }
+
+  if (invalidMiddleware.length > 0) {
+    console.error(
+      `Next.js Plugin: ${
+        pluginMetaData.name
+      } listed invalid middleware ${invalidMiddleware.join(', ')}`
+    )
+  }
+
+  // TODO: investigate requiring plugins' env be prefixed
+  // somehow to prevent collision
+  if (!Array.isArray(pluginMetaData['required-env'])) {
+    exitWithError(
+      'Next.js plugins need to have a "nextjs.required-env" key in package.json'
+    )
+  }
+
+  const missingEnvFields: string[] = []
+
+  for (const field of pluginMetaData['required-env']) {
+    if (typeof env[field] === 'undefined') {
+      missingEnvFields.push(field)
+    }
+  }
+
+  if (missingEnvFields.length > 0) {
+    exitWithError(
+      `Next.js Plugin: ${
+        pluginMetaData.name
+      } required env ${missingEnvFields.join(
+        ', '
+      )} but was missing in your \`next.config.js\``
+    )
+  }
+
+  return {
+    middleware,
+    directory: pkgDir,
+    requiredEnv: pluginMetaData['required-env'],
+    pluginName: pluginMetaData.name,
+    pkgName: pluginPackageJson.name,
+  }
+}
+
+type SeparatedPlugins = {
+  appMiddlewarePlugins: PluginMetaData[]
+  documentMiddlewarePlugins: PluginMetaData[]
+}
+
+// clean package name so it can be used as variable
+export const getPluginId = (pkg: string): string => {
+  pkg = pkg.replace(/\W/g, '')
+
+  if (pkg.match(/^[0-9]/)) {
+    pkg = `_${pkg}`
+  }
+  return pkg
+}
+
+export function getSeparatedPlugins(
+  plugins: PluginMetaData[]
+): SeparatedPlugins {
+  const appMiddlewarePlugins = []
+  const documentMiddlewarePlugins = []
+
+  for (const plugin of plugins) {
+    let addedFor_app = false
+    let addedFor_document = false
+
+    // TODO: add checking if valid middleware export
+    for (const middleware of plugin.middleware) {
+      if (!addedFor_app && middleware.startsWith('_app.')) {
+        appMiddlewarePlugins.push(plugin)
+        addedFor_app = true
+      }
+      if (!addedFor_document && middleware.startsWith('_document.')) {
+        documentMiddlewarePlugins.push(plugin)
+        addedFor_document = true
+      }
+    }
+  }
+
+  return {
+    appMiddlewarePlugins,
+    documentMiddlewarePlugins,
+  }
+}
+
+export async function collectPlugins(
+  dir: string,
+  env: ENV_OPTIONS
+): Promise<PluginMetaData[]> {
+  const rootPackageJsonPath = await findUp('package.json', { cwd: dir })
+  if (!rootPackageJsonPath) {
+    return []
+  }
+  const rootPackageJson = require(rootPackageJsonPath)
+
+  let dependencies: string[] = []
+  if (rootPackageJson.dependencies) {
+    dependencies = dependencies.concat(
+      Object.keys(rootPackageJson.dependencies)
+    )
+  }
+
+  if (rootPackageJson.devDependencies) {
+    dependencies = dependencies.concat(
+      Object.keys(rootPackageJson.devDependencies)
+    )
+  }
+
+  const nextPluginNames = dependencies.filter(name => {
+    return name.startsWith('next-plugin-') || name.startsWith('@next/plugin-')
```

Comment:
This should also take scoped packages into account.

Reply:
Updated